### PR TITLE
feat(platform-server): provide a DOM implementation on the server

### LINKED ISSUE #14638: Provide DOM interface to DOCUMENT on server side
This is for the 4.1 timeframe.

Currently the DOCUMENT you get in the browser is an actual 'document', while in the server it is a parse5 Document.

We need to get rid of the DOM Adapter(and salvage only the helper functions) and provide an actual DOM interface to the DOCUMENT returned on the server side (using jsdom?)

This would make the DOCUMENT actually useful without conditional logic on the browser and server.

### PR DESCRIPTION
**Pull request by vikerman**
Fixes #14638

Uses Domino - https://github.com/fgnass/domino and removes dependency on
Parse5.

The DOCUMENT and nativeElement were never typed earlier and were
different on the browser(DOM nodes) and the server(Parse5 nodes). With
this change, platform-server also exposes a DOCUMENT and nativeElement
that is closer to the client.

Removes the need to add services for each and every Document
manipulation like Title/Meta etc.

This does *not* provide a global variable 'document' or 'window' on the
server. You still have to inject DOCUMENT to get the document backing
the current platform server instance.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[x] Yes
[] No
```
If previously your code was manipulating the document directly instead of using the renderer, you would have to migrate that code from using the Parse5 API to using just the DOM API(the subset supported by Domino).

### CODE REVIEW

File: packages/platform-browser/testing/src/browser_util.ts [hunk 1]
```diff
@@ -72,12 +72,13 @@ export class BrowserDetection {
 
 BrowserDetection.setup();
 
-export function dispatchEvent(element: any, eventType: any): void {
-  getDOM().dispatchEvent(element, getDOM().createEvent(eventType));
+export function dispatchEvent(element: any, eventType: any, defaultDoc: Document): void {
+  getDOM().dispatchEvent(element, getDOM().createEvent(eventType, defaultDoc));
```
Comment (mhevery) [inline packages/platform-browser/testing/src/browser_util.ts]: `element.ownerDocument`

File: packages/platform-browser/testing/src/browser_util.ts [hunk 2]
```diff
@@ -72,12 +72,13 @@ export class BrowserDetection {
...
 }
 
-export function el(html: string): HTMLElement {
-  return <HTMLElement>getDOM().firstChild(getDOM().content(getDOM().createTemplate(html)));
+export function el(html: string, defaultDoc: Document): HTMLElement {
+  return <HTMLElement>getDOM().firstChild(
+      getDOM().content(getDOM().createTemplate(html, defaultDoc)));
```
Comment (mhevery) [inline packages/platform-browser/testing/src/browser_util.ts]: `getDOM().getDefaultDoc`

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/platform-browser/test/animation/animation_renderer_spec.ts [hunk 3]
```diff
@@ -109,160 +111,159 @@ export function main() {
...
-          expect(cmp.event.triggerName).toEqual('myAnimation');
-          expect(cmp.event.phaseName).toEqual('start');
-          cmp.event = null;
-
-          engine.flush();
-          expect(cmp.event).toBeFalsy();
-          async();
-        });
-      });
+  describe('flushing animations', () => {
```
Comment (mhevery) [inline packages/platform-browser/test/animation/animation_renderer_spec.ts]: find out what is going on here with all of the removals and additions.

File: packages/platform-server/src/domino_adapter.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  getDefaultDocument(): Document { return defaultDoc !; }
+
+  createShadowRoot(el: any, doc: Document = document): DocumentFragment {
+    el.shadowRoot = doc.createDocumentFragment();
+    el.shadowRoot.parent = el;
+    return el.shadowRoot;
+  }
+  getShadowRoot(el: any): DocumentFragment { return el.shadowRoot; }
+
+  isTextNode(node: any): boolean { return node.nodeType === TEXT_NODE; }
```
Comment (mhevery) [inline packages/platform-server/src/domino_adapter.ts]: Can we use `Document.TEXT_NODE`

File: packages/platform-server/src/domino_adapter.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  isCommentNode(node: any): boolean { return node.nodeType === COMMENT_NODE; }
```
Comment (mhevery) [inline packages/platform-server/src/domino_adapter.ts]: Can we use `Document.COMMENT_NODE`

File: packages/platform-server/src/domino_adapter.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  isElementNode(node: any): boolean { return node ? node.nodeType === ELEMENT_NODE : false; }
```
Comment (mhevery) [inline packages/platform-server/src/domino_adapter.ts]: Can we use `Document.ELEMENT_NODE`

File: packages/platform-server/src/domino_adapter.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+const domino = require('domino');
+
+import {ɵBrowserDomAdapter as BrowserDomAdapter, ɵsetRootDomAdapter as setRootDomAdapter} from '@angular/platform-browser';
+
+const TEXT_NODE = 3;
```
Comment (mhevery) [inline packages/platform-server/src/domino_adapter.ts]: Remove the constants

File: packages/platform-server/src/domino_adapter.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+import {ɵBrowserDomAdapter as BrowserDomAdapter, ɵsetRootDomAdapter as setRootDomAdapter} from '@angular/platform-browser';
+
+let Document = domino.createDOMImplementation().createDocumentType();
+
+function _notImplemented(methodName: string) {
+  return new Error('This method is not implemented in DominoAdapter: ' + methodName);
+}
+
+let defaultDoc: Document|null = null;
```
Comment (gdi2290) [inline packages/platform-server/src/domino_adapter.ts]: we can't do this without running into concurrency issues (two requests with async http calls same document)

Comment (vikerman) [inline packages/platform-server/src/domino_adapter.ts]: Used only for tests.

**Review by mhevery — APPROVED** [review]

**Comment by jasonaden** [comment]
Merged

**Pull request closed without merging**